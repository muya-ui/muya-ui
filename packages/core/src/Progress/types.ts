import { ISvgProps } from '@muya-ui/theme-light';

import { CSSProperties } from 'react';

import { Omit } from '../types';

export interface IProgressProps extends Omit<ISvgProps, 'children'> {
  /**
   * 进度条类型
   * @default 'line'
   */
  type?: 'line' | 'circle';
  /**
   * 进度条状态
   *
   * @type {('normal' | 'success' | 'error')}
   * @memberof IProgressProps
   * @default 'normal'
   */
  status?: 'normal' | 'success' | 'error';
  /**
   * 进度条的百分比值
   *
   * @type {number}
   * @memberof IProgressProps
   * @default 0
   */
  progress?: number;
  /**
   * 进度条主体的颜色
   *
   * @type {string}
   * @memberof IProgressProps
   */
  progressColor?: string;
  /**
   * 进度条背景的颜色
   *
   * @type {string}
   * @memberof IProgressProps
   */
  bgColor?: string;
  /**
   * 文字的颜色
   *
   * @type {string}
   * @memberof IProgressProps
   */
  textColor?: string;
  /**
   * 是否需要动画效果
   *
   * @type {boolean}
   * @memberof IProgressProps
   * @default true
   */
  animate?: boolean;
  /**
   * 动画持续时间
   *
   * @type {number}
   * @memberof IProgressProps
   */
  animationDuration?: number;
  /**
   * 是否展示百分比文案
   *
   * @type {boolean}
   * @memberof IProgressProps
   * @default true
   */
  showPercentage?: boolean;
  /**
   * 是否展示 "%"
   *
   * @type {boolean}
   * @memberof IProgressProps
   * @default true
   */
  showPercentageSymbol?: boolean;
  /**
   * 文案与"%"之间的距离
   *
   * @type {number}
   * @memberof IProgressProps
   * @default 0
   */
  percentSpacing?: number;
  /**
   * 百分比文字节点的样式
   *
   * @type {CSSProperties}
   * @memberof IProgressProps
   */
  textStyle?: CSSProperties;
  /**
   * 自定义类名
   *
   * @type {string}
   * @memberof IProgressProps
   */
  className?: string;
  /**
   * 自定义样式
   *
   * @type {CSSProperties}
   * @memberof IProgressProps
   */
  style?: CSSProperties;
  /**
   * 环状进度条的宽度，只能传入百分比数值
   * @default 12.5
   * @type {string}
   * @memberof IProgressProps
   */
  circleLineWidth?: number;
  /**
   * 环状进度条两端的形状是否为圆形
   *
   * @default true
   * @type {boolean}
   * @memberof IProgressProps
   */
  roundedStroke?: boolean;
}
