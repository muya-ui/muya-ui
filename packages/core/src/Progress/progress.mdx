---
name: Progress 进度条
menu: 组件
route: /components/progress
autoDemos: true
parent: 反馈
---

# Progress 进度条

展示操作的当前进度。

## 使用场景

- 在操作需要较长时间才能完成时，为用户显示该操作的当前进度和状态。
- 当一个操作会打断当前界面，或者需要在后台运行，且耗时可能超过 2 秒时。
- 当需要显示一个操作完成的百分比时。

{Demos}

### API

#### IProgressProps

<Props propsName="IProgressProps" />
