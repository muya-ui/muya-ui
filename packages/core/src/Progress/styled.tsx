import styled, { css } from 'styled-components';

export interface IStyledProgressLineProps {
  $bgColor: string;
}

export const StyledProgressLine = styled.div<IStyledProgressLineProps>`
  ${props => css`
    width: 100%;
    height: ${props.theme.components.Progress.height}px;
    background: ${props.$bgColor};
    border-radius: ${props.theme.components.Progress.borderRadius};

    ${StyledProgressBg} {
      height: 100%;
      border-radius: ${props.theme.components.Progress.borderRadius};
    }
  `}
`;

export const StyledProgressBg = styled.div``;
