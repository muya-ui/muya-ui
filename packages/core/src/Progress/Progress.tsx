import React, { Ref, useMemo } from 'react';

import { SvgIcon } from '@muya-ui/theme-light';

import useTheme from '../utils/useTheme';
import { StyledProgressBg, StyledProgressLine } from './styled';
import { IProgressProps } from './types';
import memoForwardRef from '../utils/memoForwardRef';

export default memoForwardRef((props: IProgressProps, ref: Ref<SVGSVGElement>) => {
  const { status = 'normal', progress = 0 } = props;
  const theme = useTheme();
  const {
    transition: { pattern: transitionPattern },
    colors: { pattern: colorsPattern },
    components: { Progress: token },
  } = theme;
  let defaultProgressColor = colorsPattern.feature.info;
  let defaultTextColor = colorsPattern.text.assistant;

  if (status === 'error') {
    defaultProgressColor = colorsPattern.feature.error;
  } else if (status === 'success' || progress >= 100) {
    defaultTextColor = colorsPattern.feature.success;
    defaultProgressColor = colorsPattern.feature.success;
  }

  const {
    // color
    textStyle,
    textColor = defaultTextColor,
    progressColor = defaultProgressColor,
    bgColor = token.defaultBgColor,

    // precent text
    percentSpacing = 0,
    showPercentage = true,
    showPercentageSymbol = true,

    // animate
    animate = true,
    animationDuration = transitionPattern.duration.status,

    fontSize = token.defaultFontSize,
    style,
    className,
    type = 'line',
    circleLineWidth: circleWidthPercent = 12.5,
    roundedStroke = true,
    ...other
  } = props;

  const outerRadius = 200;
  const circleLineWidth = (circleWidthPercent * outerRadius) / 100;
  const radius = outerRadius - circleLineWidth;
  const diameter = Math.round(Math.PI * radius * 2); // 内部圆直径
  const strokeDashoffset = Math.round(((100 - Math.min(progress, 100)) / 100) * diameter);
  const transition = animate
    ? `${type === 'line' ? 'width' : 'stroke-dashoffset'} ${animationDuration}ms ${
        transitionPattern.easing.status
      }`
    : undefined;
  const viewBox = `-${circleLineWidth} -${circleLineWidth} ${outerRadius * 2} ${outerRadius * 2}`;

  const textElement = useMemo(
    () => (
      <text
        style={textStyle}
        fill={textColor}
        x={radius}
        y={radius}
        fontSize={radius / 2}
        textAnchor="middle"
        dominantBaseline="central"
      >
        <tspan textAnchor="middle" dominantBaseline="central">
          {progress}
        </tspan>
        {showPercentageSymbol && (
          <tspan textAnchor="middle" dominantBaseline="central" dx={percentSpacing}>
            %
          </tspan>
        )}
      </text>
    ),
    [percentSpacing, progress, radius, showPercentageSymbol, textColor, textStyle],
  );

  if (type === 'line') {
    return (
      <StyledProgressLine theme={theme} $bgColor={bgColor} style={style} className={className}>
        <StyledProgressBg
          style={{
            width: `${progress}%`,
            background: progressColor,
            transition,
          }}
        />
      </StyledProgressLine>
    );
  }
  return (
    <SvgIcon
      ref={ref}
      style={style}
      className={className}
      fontSize={fontSize}
      viewBox={viewBox}
      {...other}
    >
      <circle
        stroke={bgColor}
        cx={radius}
        cy={radius}
        r={radius}
        strokeWidth={circleLineWidth}
        fill="none"
      />
      <circle
        stroke={progressColor}
        transform={`rotate(-90 ${radius} ${radius})`}
        cx={radius}
        cy={radius}
        r={radius}
        strokeDasharray={diameter}
        strokeDashoffset={diameter}
        strokeWidth={circleLineWidth}
        strokeLinecap={roundedStroke ? 'round' : 'butt'}
        fill="none"
        style={{ strokeDashoffset, transition }}
      />
      {showPercentage ? textElement : <></>}
    </SvgIcon>
  );
});
