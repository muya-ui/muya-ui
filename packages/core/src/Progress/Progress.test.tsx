import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Progress from './Progress';

test('should render correctly', () => {
  const tree = renderer
    .create(
      <>
        <Progress />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('should accept status', () => {
  const tree = renderer
    .create(
      <>
        <Progress status="success" progress={10} />
        <Progress status="error" progress={10} />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('should disable animate', () => {
  const tree = renderer.create(<Progress progress={10} animate={false} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should be circle Progress', () => {
  const wrapper = mount(<Progress progress={10} type="circle" />);
  expect(wrapper.find('svg')).toHaveLength(1);
});

test('should hide text', () => {
  const wrapper = mount(
    <Progress progress={10} roundedStroke={false} type="circle" showPercentage={false} />,
  );
  expect(wrapper.find('text')).toHaveLength(0);
});
