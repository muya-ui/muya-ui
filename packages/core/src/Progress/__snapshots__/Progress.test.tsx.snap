// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept status 1`] = `
Array [
  .c0 {
  width: 100%;
  height: 6px;
  background: rgba(51,51,51,0.04);
  border-radius: 999rem;
}

.c0 .c1 {
  height: 100%;
  border-radius: 999rem;
}

<div
    className="c0"
  >
    <div
      className="c1 "
      style={
        Object {
          "background": "rgba(0,217,72,1)",
          "transition": "width 150ms cubic-bezier(0.4, 0, 0.2, 1)",
          "width": "10%",
        }
      }
    />
  </div>,
  .c0 {
  width: 100%;
  height: 6px;
  background: rgba(51,51,51,0.04);
  border-radius: 999rem;
}

.c0 .c1 {
  height: 100%;
  border-radius: 999rem;
}

<div
    className="c0"
  >
    <div
      className="c1 "
      style={
        Object {
          "background": "rgba(255,43,0,1)",
          "transition": "width 150ms cubic-bezier(0.4, 0, 0.2, 1)",
          "width": "10%",
        }
      }
    />
  </div>,
]
`;

exports[`should disable animate 1`] = `
.c0 {
  width: 100%;
  height: 6px;
  background: rgba(51,51,51,0.04);
  border-radius: 999rem;
}

.c0 .c1 {
  height: 100%;
  border-radius: 999rem;
}

<div
  className="c0"
>
  <div
    className="c1 "
    style={
      Object {
        "background": "rgba(26,122,248,1)",
        "transition": undefined,
        "width": "10%",
      }
    }
  />
</div>
`;

exports[`should render correctly 1`] = `
.c0 {
  width: 100%;
  height: 6px;
  background: rgba(51,51,51,0.04);
  border-radius: 999rem;
}

.c0 .c1 {
  height: 100%;
  border-radius: 999rem;
}

<div
  className="c0"
>
  <div
    className="c1 "
    style={
      Object {
        "background": "rgba(26,122,248,1)",
        "transition": "width 150ms cubic-bezier(0.4, 0, 0.2, 1)",
        "width": "0%",
      }
    }
  />
</div>
`;
